tests/test_nvfuser.py

```diff
@@ -0,0 +1,172 @@
+
+try:
+    from xformers.components.nvfuser import (
+        NVFusedBiasActivationDropout,
+        NVFusedBiasDropoutRes,
+        NVFusedBiasDropoutResLayerNorm,
+    )
+    from xformers.components.nvfuser.utils import build_nvfused
+except ImportError as e:
+    logging.warning(f"Functorch is not available in test_nvfuser.py. \nError {e}")
+
+FUSED_PATTERNS = (
+    [
+        NVFusedBiasActivationDropout,
+        NVFusedBiasDropoutRes,
+        NVFusedBiasDropoutResLayerNorm,
+    ]
+    if xformers._is_functorch_available
+    else []
+)
+
+# Testing odd (non-power-of-two for instance) shapes on purpose
+SHAPES = [
+    (384, 512),
+    (8, 384, 128),
+    (8, 784, 512),
+    (4, 16, 384),
+    (4, 16, 1024),
+    (2, 16, 2048),
+    (2, 16, 4096),
+    (1, 16, 12288),
+]
+
+BATCH = 4
+SEQ = 256
+EMBD = 16
+LATENT = 128
+DEVICES = [torch.device("cuda")]
+
+ACTIVATIONS = [
+    Activation.ReLU,
+    Activation.GeLU,
+    Activation.LeakyReLU,
+    Activation.SquaredReLU,
+    Activation.SmeLU,
+]
+
+
+@pytest.mark.skipif(
+    not xformers._is_functorch_available, reason="Functorch is not available"
+)
+@pytest.mark.skipif(not _gpu_available, reason="GPU is not available")
+@pytest.mark.parametrize("fused_pattern", FUSED_PATTERNS)
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("amp", [False, True])
+@pytest.mark.parametrize("bias", [False, True])
+@pytest.mark.parametrize("activation", ACTIVATIONS)
+@pytest.mark.parametrize("p", [0, 0.1, 0.5])
+@pytest.mark.parametrize("layer_norm_style", [LayerNormStyle.Pre, LayerNormStyle.Post])
+def test_nvfused_pattern_parity(
+    fused_pattern: nn.Module,
+    shape: tuple,
+    amp: bool,
+    bias: bool,
+    activation: Activation,
+    p: float,
+    layer_norm_style: LayerNormStyle,
+):
+
+    if (
+        fused_pattern != NVFusedBiasDropoutResLayerNorm
+        and layer_norm_style == LayerNormStyle.Post
+    ):
+        pytest.skip(
+            "Layer norm style doesn't apply, the same relevant params already tested once."
+        )
+
+    torch.cuda.manual_seed_all(0)
+    torch.random.manual_seed(0)
+    x = torch.normal(0, 1, size=shape, device="cuda", requires_grad=True)
+    x_cpu = x.clone().cpu()
+
+    with autocast(enabled=amp):
+        fused = build_nvfused(
+            fused_pattern, shape, bias, activation, p, layer_norm_style
+        )
+        fused.train().cuda()
+        nvfused_res = fused(x, x) if fused.requires_residual else fused(x)
+        fused.cpu()
+        torch_res = (
+            fused(x_cpu, x_cpu).cuda()
+            if fused.requires_residual
+            else fused(x_cpu).cuda()
+        )
+
+        # Check if operation was actually fused
+        assert isinstance(
+            nvfused_res.grad_fn, torch.autograd.function.BackwardCFunction
+        )
+
+        if p == 0.0:
+            # Check fused and unfused paths are the same
+            assert torch.allclose(torch_res, nvfused_res, atol=1e-6, rtol=1e-2)
```

Comment:
does `torch.equal` fail?

Reply:
Yes it does, for some reason

Reply:
Does it also fail for smaller atol and rtol?